packages/bem/src/types.ts

```diff
@@ -0,0 +1,260 @@
+  : never
+
+type ApplyStyleToModifierPrefix<style extends BemStyle, r extends string> = style extends 'classic'
+  ? `${r}_`
+  : style extends 'two-dashes'
+  ? `${r}--`
+  : never
+
+type ApplyStyleToElementPrefix<style extends BemStyle, r extends string> = `${r}__`
+
+type ReadonlyStringArray = readonly string[]
+
+export interface BlockBuilder<
+  in out r extends string,
+  in out mm extends ModifiersArray = [],
+  in out ee extends BuiltElem<any, any>[] = [],
+> {
+  __virtual: [r, mm, ee]
+  mod: {
+    <m extends string>(key: m): BlockBuilder<r, AddModifier<mm, m>, ee>
+    <m extends string, vv extends ReadonlyStringArray>(key: m, values: vv): BlockBuilder<r, AddModifier<mm, m, vv>, ee>
+    <m extends string, v extends string>(key: m, value: v): BlockBuilder<r, AddModifier<mm, m, v>, ee>
+  }
+  elem: {
+    <n extends string>(name: n): BlockBuilder<r, mm, [...ee, BuiltElem<n, []>]>
+    <n extends string, e extends ElemBuilder<any>>(name: n, fn: (elementBuilder: ElemBuilder) => e): BlockBuilder<
+      r,
+      mm,
+      [...ee, BuildElement<n, e>]
+    >
+  }
+  build: {
+    (): BuildClasses<r, mm, ee, 'classic'>
+    <s extends BemStyle>(style: s): BuildClasses<r, mm, ee, s>
+  }
+}
+
+type AnyBlockBuilder = BlockBuilder<any, any, any>
+
+type ExtendBlockWithElement<b extends AnyBlockBuilder, e extends BuiltElem<any, any>> = b extends BlockBuilder<
+  infer r,
+  infer m,
+  infer ee
+>
+  ? BlockBuilder<r, m, [...ee, e]>
+  : never
+
+type BuildElement<n extends string, b extends ElemBuilder<any>> = b extends ElemBuilder<infer m>
+  ? BuiltElem<n, m>
+  : never
+
+interface ElemBuilder<in out mm extends ModifiersArray = []> {
+  __virtual: mm
+  mod: {
+    <k extends string>(key: k): ElemBuilder<AddModifier<mm, k>>
+    <k extends string, vv extends ReadonlyStringArray>(key: k, values: vv): ElemBuilder<AddModifier<mm, k, vv>>
+    <k extends string, v extends string>(key: k, value: v): ElemBuilder<AddModifier<mm, k, v>>
+  }
+}
+
+type AnyElemBuilder = ElemBuilder<any>
+
+interface BuiltElem<n extends string, m extends ModifiersArray> {
+  name: n
+  modifiers: m
+}
+
+interface ModifierKeyValue<M extends string, V extends string> {
+  kind: 'modifier-k-v'
+  key: M
+  value: V
+}
+
+interface ModifierBool<M extends string> {
+  kind: 'modifier-bool'
+  key: M
+}
+
+type AnyModifier = ModifierBool<any> | ModifierKeyValue<any, any>
+
+type ModifiersArray = Array<ModifierKeyValue<any, any> | ModifierBool<any>>
+
+type ExpandModifiersTuple<k extends string, vv extends ReadonlyStringArray> = vv extends readonly [
+  infer head extends string,
+  ...infer tail extends ReadonlyStringArray,
+]
+  ? [ModifierKeyValue<k, head>, ...ExpandModifiersTuple<k, tail>]
+  : []
+
+type test14 = Expect<
+  Equal<ExpandModifiersTuple<'key', ['a', 'b']>, [ModifierKeyValue<'key', 'a'>, ModifierKeyValue<'key', 'b'>]>
+>
+
+type AddModifier<
+  mm extends ModifiersArray,
+  k extends string,
+  v extends ReadonlyStringArray | string | null = null,
```

Comment:
Please give readable names to type valiables like `Value extends ReadonlyStringArray | string | null`

Reply:
Okay.